### ISSUE TITLE
Invalid transformation to const function with deref

### ISSUE THREAD

**Issue by SUPERCILEX** [rust-lang/rust-clippy#14091]
### Summary

```rust
struct BucketSlotGuard<'a> {
    id: u32,
    free_list: &'a mut Vec<u32>,
}

impl BucketSlotGuard<'_> {
    fn into_inner(self) -> u32 {
        let this = ManuallyDrop::new(self);
        this.id
    }
}
```

```
The following errors were reported:
error[E0015]: cannot perform non-const deref coercion on `std::mem::ManuallyDrop<allocator::BucketSlotGuard<'_>>` in constant functions
   --> server/src/allocator.rs:168:9
    |
168 |         this.id
    |         ^^^^^^^
    |
    = note: attempting to deref into `allocator::BucketSlotGuard<'_>`
note: deref defined here
   --> /home/asaveau/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/mem/manually_drop.rs:262:5
    |
262 |     type Target = T;
    |     ^^^^^^^^^^^
note: impl defined here, but it is not `const`
   --> /home/asaveau/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/mem/manually_drop.rs:261:1
    |
261 | impl<T: ?Sized> Deref for ManuallyDrop<T> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants

```

```
warning: this could be a `const fn`
   --> server/src/allocator.rs:166:5
    |
166 | /     fn into_inner(self) -> u32 {
167 | |         let this = ManuallyDrop::new(self);
168 | |         this.id
169 | |     }
    | |_____^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_const_for_fn
    = note: `-W clippy::missing-const-for-fn` implied by `-W clippy::nursery`
    = help: to override `-W clippy::nursery` add `#[allow(clippy::missing_const_for_fn)]`
help: make the function `const`
    |
166 |     const fn into_inner(self) -> u32 {
    |     +++++

```

### Version

```text
rustc 1.86.0-nightly (2f348cb7c 2025-01-27)
binary: rustc
commit-hash: 2f348cb7ce4063fa4eb40038e6ada3c5214717bd
commit-date: 2025-01-27
host: x86_64-unknown-linux-gnu
release: 1.86.0-nightly
LLVM version: 19.1.7
```

### Error output

<details><summary>Backtrace</summary>
  <p>

  ```
  <backtrace>
  ```

  </p>
</details>

**Comment by matthiaskrgr** [comment]
another example from rustc
```rust
// Regression test for #30104

//@ check-pass
#![allow(unused)]

use std::ops::{Deref, DerefMut};

fn box_two_field(v: &mut Box<(i32, i32)>) {
    let _a = &mut v.0;
    let _b = &mut v.1;
}

fn box_destructure(v: &mut Box<(i32, i32)>) {
    let (ref mut _head, ref mut _tail) = **v;
}

struct Wrap<T>(T);

impl<T> Deref for Wrap<T> {
    type Target = T;
    fn deref(&self) -> &T {
        &self.0
    }
}

impl<T> DerefMut for Wrap<T> {
    fn deref_mut(&mut self) -> &mut T {
        &mut self.0
    }
}

fn smart_two_field(v: &mut Wrap<(i32, i32)>) {
    let _a = &mut v.0;
    let _b = &mut v.1;
}

fn smart_destructure(v: &mut Wrap<(i32, i32)>) {
    let (ref mut _head, ref mut _tail) = **v;
}

fn main() {}

```


```
warning: this could be a `const fn`
  --> ./tests/ui/nll/issue-30104.rs:32:1
   |
32 | / fn smart_two_field(v: &mut Wrap<(i32, i32)>) {
33 | |     let _a = &mut v.0;
34 | |     let _b = &mut v.1;
35 | | }
   | |_^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_const_for_fn
   = note: requested on the command line with `--force-warn clippy::missing-const-for-fn`
help: make the function `const`
   |
32 | const fn smart_two_field(v: &mut Wrap<(i32, i32)>) {
   | +++++

warning: this could be a `const fn`
  --> ./tests/ui/nll/issue-30104.rs:37:1
   |
37 | / fn smart_destructure(v: &mut Wrap<(i32, i32)>) {
38 | |     let (ref mut _head, ref mut _tail) = **v;
39 | | }
   | |_^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_const_for_fn
help: make the function `const`
   |
37 | const fn smart_destructure(v: &mut Wrap<(i32, i32)>) {
   | +++++
```

```
error[E0015]: cannot call non-const method `<Wrap<(i32, i32)> as std::ops::DerefMut>::deref_mut` in constant functions
  --> ./tests/ui/nll/issue-30104.rs:34:19
   |
34 |     let _b = &mut v.1;
   |                   ^^^
   |
   = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants

error[E0015]: cannot call non-const method `<Wrap<(i32, i32)> as std::ops::DerefMut>::deref_mut` in constant functions
  --> ./tests/ui/nll/issue-30104.rs:38:42
   |
38 |     let (ref mut _head, ref mut _tail) = **v;
   |                                          ^^^
   |
   = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0015`.
```

**Comment by profetia** [comment]
Should also be fixed in #14294.
